tests/unit/test_awsrequest.py

```diff
@@ -134,6 +134,26 @@ def test_can_prepare_url_params_with_existing_query(self):
             prepared_request.url, 'http://example.com/?bar=foo&foo=bar'
         )
 
+    def test_can_use_list_tuples_for_params(self):
+        request = AWSRequest(
+            url='http://example.com/', params=[('foo', 'bar')]
+        )
+        prepared_request = request.prepare()
+        self.assertEqual(
+            prepared_request.url, 'http://example.com/?foo=bar'
+        )
+
+    def test_can_use_list_duplicate_tuples_for_params(self):
```

Comment:
We can probably use `pytest.parametrize` to collapse this into a single test with some more cases. Off the top of my head, we'll probably want to evaluate cases like:

* `[("foo", None)]`
* `[(,)]`
* `[(None, None)]`
* `[("foo", "bar"), ("foo", "baz"), ("fizz", "buzz")]`
* `[("food", "bar"), ("foo", None)]`

Reply:
Oh cool! Didn't know about that feature. Very handy. The second option you provided raises a syntax error. Should I do something like two empty strings instead or some other falsy values?

Reply:
We were looking for an empty tuple. That's likely a failure case (which may be a different test) but I wanted to make sure we understand the failure states the list is introducing. We can do empty strings too (an empty key is what we specifically care about).

Reply:
ok cool thank you for clarifying

Reply:
Also as a note, you'll need to move this out of the test class to use the `pytest.mark.parametrize` decorator. You can find an example [here](https://github.com/boto/botocore/blob/develop/tests/functional/test_model_completeness.py#L28-L31). You also don't need that generator function, you can write the cases inline like:

```python
@pytest.mark.parametrize(
    "params, expected_url",
    ([("foo", None)], "...?foo"),
    ([()], "...<no query string?>"),
    ...
)
```

In general we should be moving in that direction, the current testing layouts are due to our former test runner, `nosetest`.